
*Klebsiella pneumoniae* (\ *Kp*\ ) is a commensal bacterium that causes opportunistic infections in hospitals. It has six close relatives (species and subspecies), and together these are known as the  *K. pneumoniae* species complex (KpSC) <https://github.com/katholt/Kleborate/wiki/Species-detection#k-pneumoniae-species-complex-kpsc>`_. These related species are often difficult to distinguish from one another in clinical labs using biotyping or MALDI-TOF and consequently can be confused for *K. pneumoniae sensu stricto*.

*K. pneumoniae* are intrinsically resistant to ampicillin, and additional antimicrobial resistances frequently arise through horizontal gene transfer and/or chromosomal mutations. Multi-drug resistance (MDR) is increasing globally and MDR strains with resistance to the carbapenems are of particular concern, earning *K. pneumoniae* top position in the World Health Organization's priority list of antimicrobial resistant pathogens for which novel control strategies are urgently required. A handful of 'hypervirulent' *K. pneumoniae* clonal groups are also recognised, comprising strains that encode a constellation of acquired virulence factors and which can cause invasive disease outside the hospital setting. Fortunately, most of these hypervirulent strains have so far remained susceptible to the majority of antimicrobials. But evidence is now mounting that other *K. pneumoniae*\ , including MDR and carbapenem-resistant strains, can acquire the virulence factors – siderophores (yersiniabactin, salmochelin and aerobactin), regulators of hypermucoidy (the rmpADC locus and potentially also rmpA2) and/or the genotoxin colibactin – resulting in enhanced virulence potential. This so-called 'convergence' of MDR and acquired virulence factors further heightens the public health risk associated with *K. pneumoniae* because the resulting strains have the potential to cause severe infections that are extremely difficult to treat. In this regard, capsule (K) and LPS (O) antigen variation is of increasing interest to the research community due to its importance in host-pathogen and phage interactions, and thus potential relevance to novel *K. pneumoniae* control measures such as vaccines, immunotherapy and phage therapy. 

To learn more about taxonomy and population genomics of *Klebsiella pneumoniae* and the KpSC, and what we know so far about the distribution of AMR, virulence, K and O types in the *K. pneumoniae* population, see `Wyres, Lam & Holt, 2020, Nature Reviews Microbiology <https://www.nature.com/articles/s41579-019-0315-1>`_.

To make it easier to extract clinically relevant genotyping information on *K. pneumoniae* and the KpSC using genome data, we have developed Kleborate- a genomic surveillance tool designed to: 


* accurately identify species and sequence types
* identify the key *acquired* genetic features for which there is strong evidence of association with either antibiotic resistance or hypervirulence in *K. pneumoniae sensu stricto*
* optionally predict K and O types

While many generic tools can be used to identify sequence types or resistance determinants from bacterial genomes, we hope that this organism-specific tool will help avoid many of the common confusions faced by people working with *K. pneumoniae* genomes, and also facilitate monitoring for the convergence of antimicrobial resistance with the hypervirulence-associated factors noted above.
